### PR TITLE
Close #6762: latex: Allow to load additonal LaTeX packages

### LINKED ISSUE #6762: Documentation for <packages> field in latex.tex_t
**Problem description leading to this feature request**
When using `make latexpdf` (or any latex-related), it may be interesting to add extra packages to the preamble.

However, when adding packages via the `preamble` field of the `latex_elements`, they are added after the `hyperref` package.

This may lead to unwanted results with the 
```
\usepackage[makeindex]{imakeidx}
\usepackage[columns=1]{idxlayout}
```
for instance. Indeed, these commands will be added after the `hyperref` package, resulting in an incomplete `.ind` file (instead of `\hyperpage{9}` one would only get `9`).

**Is their a workaround? Yes.**
The workaround I use for the moment is to add
```
\usepackage[makeindex]{imakeidx}
\usepackage[columns=1]{idxlayout}
```
in the `passoptionstopackages` field of `latex_elements`.

**Proposed solution**
I looked at [https://github.com/sphinx-doc/sphinx/blob/master/sphinx/templates/latex/latex.tex_t](url), which is the template file I supposed is used by Sphinx. There is (lines 39-45)
```
<%- for name, option in packages %>
<%- if option %>
\usepackage[<%= option %>]{<%= name %>}
<%- else %>
\usepackage{<%= name %>}
<%- endif %>
<%- endfor %>
```
so I tried the key `packages` in the `latex_elements`, resulting in:
> WARNING: Unknown configure key: latex_elements['packages']. ignored.
Solution:
1) Maybe I miss something and it is possible to add packages before `hyperref` with others means (but after reading [https://www.sphinx-doc.org/en/master/latex.html#latex-macros-and-environments](url) I could not find any)
2) Provide a `packages` field that inserts packages where `<%- for name, option in packages %>` is located in the template. That would be great !

**The last word**
Anyway, Sphinx is an amazing tool !

### PR DESCRIPTION
**Pull request by tk0miya**
### Feature or Bugfix
- Feature

### Purpose
- refs: #6762

### CODE REVIEW

File: doc/latex.rst [hunk 1]
```diff
@@ -226,6 +226,21 @@ into the generated ``.tex`` files.  Its ``'sphinxsetup'`` key is described
...
+     package options.  For example:
+
+     .. code-block:: python
+
+         latex_elements = {
+             'packages': [('isodate', 'english'),  # => \usepackage[english]{isodate}
+                          ('xcomment', None)],     # => \usepackage{xcomment}
+         }
+
+     It defaults to empty.
```
Comment (jfbu) [inline doc/latex.rst:238]: Perhaps say the package loading is inserted after Sphinx own LaTeX style file and all packages it loads except for `hyperref` and  packages loaded from Sphinx extensions? and that packages which need to be loaded after the latter should be inserted via ``'preamble'`` key?

File: sphinx/builders/latex/__init__.py [hunk 2]
```diff
@@ -176,14 +176,15 @@ def init_context(self) -> None:
...
-        self.context['packages'] = self.usepackages
-
         # Apply user settings to context
         self.context.update(self.config.latex_elements)
         self.context['release'] = self.config.release
         self.context['use_xindy'] = self.config.latex_use_xindy
 
+        # Merge extensions' LaTeX package list to user settings
+        self.context['packages'] = (list(self.context['packages']) +
+                                    self.app.registry.latex_packages)
```
Comment (jfbu) [inline sphinx/builders/latex/__init__.py]: I see user specified packages will come first. Maybe document this (see previous comment)

File: sphinx/builders/latex/__init__.py [hunk 3]
```diff
@@ -411,6 +412,15 @@ def validate_config_values(app: Sphinx, config: Config) -> None:
             logger.warning(msg % (key,))
             config.latex_elements.pop(key)
 
+    # latex_elements['package']
+    packages = config.latex_elements.get('packages', [])
+    for pkginfo in list(packages):
+        if not isinstance(pkginfo, (list, tuple)) or len(pkginfo) != 2:
+            msg = __("latex_elements['packages'] contains should be "
+                     "a pair of package name and options: %r")
```
Comment (jfbu) [inline sphinx/builders/latex/__init__.py]: Maybe `contents` in place of `contains`? and then "should be a list of pair of package name and options"

**Review by jfbu — COMMENTED** [review]
LGTM with nits

I wonder if it will not be simpler rather than overloading the current `'packages'` which gathers its contents from extensions to add a `extrapackages` key for the latex template which would be documented as allowing to insert usage of extra packages at an earlier location than `hyperref` and packages from extension, and that the `'preamble'` key can be used for packages needed to be loaded last.

**Comment by jfbu** [comment]
> I wonder if it will not be simpler rather than overloading the current `'packages'` which gathers its contents from extensions to add a `extrapackages` key for the latex template ...

An advantage a new `latex_elements` key is that in the rare case its insertion location needs to be changed this can be achieved via user modifying latex template. This would cater for fringe case where for some reason the user extra packages must be loaded prior to hyperref (hence `'preamble'` does not fit) but after some package loaded by a Sphinx extension.

(the user could write an own extension simply to insert packages via API, but problem is that loading order relative to packages from other extensions is perhaps not easy to control robustly?)

This PR means `'packages'` key (not meant originally for user) functionality is extended but it does not make easy to specify relative location to LaTeX packages loaded via Sphinx extensions. The approach via a `extrapackages` new key would at least make this possible for user, simply by overriding LaTeX templates. (not to mention adding even one more key `'latepackages'`... which also has the problem that "late" is confusingly similar to "latex": `'lateextrapackages'`)

File: sphinx/builders/latex/__init__.py (hunk 3 above)
Comment (tk0miya) [reply sphinx/builders/latex/__init__.py]: aghhhh... its typo...

**Comment by tk0miya** [comment]
I think adding new template variable is not good way. As you said, additional variable like `late-packages` (or `extrapackage-after-extensions'` will be requested in future.

To control the order of LaTeX packages completely, it would be better to add priority to each package. And then, users can insert packages to arbitrary position::
```
latex_elements = {
    'extrapackages': [
        ('some_package', 'option', 10),   # high priority, loaded before hyperref
        ('another_pacakge', None, 200),    # lower priority, loaded after hyperref 
    ]
}
```

On the other hand, we don't get a report around the LaTeX packages added by extension. So this idea might be too much.

**Comment by jfbu** [comment]
> On the other hand, we don't get a report around the LaTeX packages added by extension. So this idea might be too much.

I believe so, too. And it is a bit unfamiliar to LaTeX world to use such "priorities". I think since it is possible for user to use own latex template, question of control of where to load a package is not an urgent one. Also  is is not easy to fix satisfactorily because perhaps complete solution means externalizing out of `sphinx.sty` all package loading to a `sphinxpackages.sty` to allow user to insert a needed own package at a precise location in a custom `sphinxpackages.sty`. (I don't say this is probable, nobody asks something like this so far...)

I feel it is simpler to have an `'extrapackages'` key for `latex_elements` with a string value directly containing various user configured `\usepackage[options]{package}`. This accommodates that for some packages usage of some `\packagenamesetup{...}`  might be better for option passing than the optional argument to `\usepackage` : for some packages usage of a "setup" macro is mandatory for some options.

Allowing such syntax would not be compatible with common treatment of ``'packages'`` key and thus reinforces legitimacy of new separate key. I think ``'packages'`` key with its Python list syntax for options has the limitation mentioned in previous paragraph and this limitation is propagated in this PR...

No strong opinion, though!

**Comment by tk0miya** [comment]
Thank you for comment. I agree your idea; to take mere string value through `extrapackages` key. I'll update this soon.

File: doc/latex.rst [hunk 4]
```diff
@@ -226,6 +226,25 @@ into the generated ``.tex`` files.  Its ``'sphinxsetup'`` key is described
...
+             'packages': r'\usepackage{isodate}'
+         }
+
+     It defaults to empty.
+
+     The specified LaTeX packages will be loaded before
+     hyperref package and packages loaded from Sphinx extensions.
+
+     .. hint:: If you'd like to load additional LaTeX packages after hyperref, use
+               ``preamble`` key instead.
```
Comment (jfbu) [inline doc/latex.rst]: perhaps ``'preamble'`` with string quotes?

**Review by jfbu — APPROVED** [review]
LGTM!

**Comment by tk0miya** [comment]
Thank you for reviewing. Merged.

**Pull request merged**